src/client_side.cc

```diff
@@ -3478,6 +3482,15 @@ clientListenerConnectionOpened(AnyP::PortCfgPointer &s, const Ipc::FdNoteId port
            << s->listenConn);
 
     Must(AddOpenedHttpSocket(s->listenConn)); // otherwise, we have received a fd we did not ask for
+
+#if USE_SYSTEMD
+    // at least one port is open, tell systemd we are able to serve connections
+    // subsequent sd_notify() calls will not be processed because first parameter is non-zero
+    // XXX: send the notification only when all configured ports are open
+    static int portsSeen = 0;
+    if (++portsSeen == NHttpSockets && (opt_foreground || opt_no_daemon))
```

Comment:
This is not my area of expertise so please feel free to dismiss the following question if the answer is obvious to an expert: Why do we need to exclude "daemonized" Squid instances? In other words, why do we need the `opt_foreground || opt_no_daemon` check here? I understand that the `opt_foreground || opt_no_daemon` check kind of matches the suggested `ExecStart` command, but I am wondering whether it is a good idea to allow _other_ ExecStart commands as well. For example, I am thinking about a custom ExecStart line that starts (in foreground, as usual) a custom script that, in turn, starts Squid (normally, i.e. in the background). AFAICT, _without_ the `opt_foreground || opt_no_daemon` check, this custom setup will work fine because Squid would still notify the grand-grand-parent systemd process.

Reply:
> This is not my area of expertise so please feel free to dismiss the following question if the answer is obvious to an expert: Why do we need to exclude "daemonized" Squid instances?

Type=notify cannot work with binaries that fork()/exit() at startup; once the parent call exit(), it is considered terminated.

> For example, I am thinking about a custom ExecStart line that starts (in foreground, as usual) a custom script that, in turn, starts Squid (normally, i.e. in the background). AFAICT, _without_ the `opt_foreground || opt_no_daemon` check, this custom setup will work fine because Squid would still notify the grand-grand-parent systemd process.

The problem is process monitoring by systemd.

I gave it a shot and, indeed, it works, only because PIDFile= is set.

But there is a race: if the shell process exits *before* daemonized Squid writes the pid file, it will fail.

Reply:
Scratch that. If the shell process does not exit, the main pid will be the shell, breaking everything (signal delivery, shutdown). This kind of setup is very fragile.

Reply:
I appreciate you investigating/explaining this issue and am marking this specific change request as addressed. Thank you.

FWIW, it feels wrong to disable notifications because simple Squid wrappers are too fragile to benefit from them. Other, more complex, wrappers could still benefit from such notifications AFAICT. However, since this is not my expertise area, and you are already spending a lot of time on critical PR changes, I am OK with ignoring this minor concern until somebody actually needs notifications from a daemonized Squid.